boot-qemu.py

```diff
@@ -0,0 +1,730 @@
+    # <kernel_dir>/.config
+    #
+    # If kernel_location is a full kernel location, it could either be:
+    #   * <kernel_dir>/.config (if the image is vmlinux)
+    #   * <kernel_dir>/../../../.config (if the image is in arch/*/boot/)
+    #   * <kernel_dir>/config (if the image is in a TuxMake folder)
+    config_file = None
+    for config_name in [".config", "../../../.config", "config"]:
+        config_path = kernel_dir.joinpath(config_name)
+        if config_path.is_file():
+            config_file = config_path.as_posix()
+            break
+
+    # Choose a sensible default value based on treewide defaults for
+    # CONFIG_NR_CPUS then get the actual value if possible.
+    config_nr_cpus = 8
+    if config_file:
+        with open(config_file) as f:
+            for line in f:
+                if "CONFIG_NR_CPUS=" in line:
+                    config_nr_cpus = int(line.split("=", 1)[1])
+                    break
+
+    # Use the minimum of the number of usable processors for the script or
+    # CONFIG_NR_CPUS.
+    usable_cpus = len(os.sched_getaffinity(0))
+    return min(usable_cpus, config_nr_cpus)
+
+
+def setup_cfg(args):
+    """
+    Sets up the global configuration based on user input.
+
+    Meaning of each key:
+
+        * append: The additional values to pass to the kernel command line.
+        * architecture: The guest architecture from the list of supported
+                        architectures.
+        * gdb: Whether or not the user wants to debug the kernel using GDB.
+        * gdb_bin: The name of or path to the GDB executable that the user
+                   wants to debug with.
+        * interactive: Whether or not the user is going to be running the
+                       machine interactively.
+        * kernel_location: The full path to the kernel image or build folder.
+        * smp_requested: Whether or not the user specified a value with
+                         '--smp'.
+        * smp_value: The value to use with '-smp' (will be used when
+                     smp_requested is True or using KVM).
+        * timeout: The value to pass along to 'timeout' if not running
+                   interactively.
+        * use_kvm: Whether or not KVM will be used.
+
+    Parameters:
+        args (Namespace): The Namespace object returned from parse_arguments()
+
+    Returns:
+        A dictionary of configuration values
+    """
+    return {
+        # Required
+        "architecture": args.architecture,
+        "kernel_location": Path(args.kernel_location).resolve().as_posix(),
+
+        # Optional
+        "append": args.append,
+        "gdb": args.gdb,
+        "gdb_bin": args.gdb_bin,
+        "interactive": args.interactive or args.gdb,
+        "smp_requested": args.smp is not None,
+        "smp_value": get_smp_value(args),
+        "timeout": args.timeout,
+        "use_kvm": can_use_kvm(args),
+    }
+
+
+def print_version_code(version):
```

Comment:
technically, it doesn't print. Rename?

Reply:
Sure: 9dc0d2048b01e39860fda2a849859c43931ba140